### ISSUE TITLE
"400 Bad Request: Token has been expired or revoked" when reading VCF via NIO

### ISSUE THREAD

**Issue by droazen** [broadinstitute/gatk#2415]
@kcibul reports the following issue while trying to use `SelectVariants` to read a VCF from GCS using the latest master:

```
I was trying this out today, and ran into some problems.  I built GATK4 from the latest master and ran:

VCF=gs://fc-4c1c7765-2de2-4214-ac41-dc10bbcbb55b/e11051c9-27ce-4ff1-ba70-de73bf11f312/PairedEndSingleSampleWorkflow/326f24b8-9158-4120-8df3-7fc4b928589c/call-GatherVCFs/S1-1.g.vcf.gz
 
./gatk-launch SelectVariants -V $VCF -L 1:1000-2000 -O foo.vcf
and I get an error about token being expired (see below).

However, I can run

gsutil cat $VCF | zcat | head
without a problem, so I don't think my gcloud credentials are the problem.

com.google.cloud.storage.StorageException: 400 Bad Request
{
  "error" : "invalid_grant",
  "error_description" : "Token has been expired or revoked."
}
	at com.google.cloud.storage.spi.DefaultStorageRpc.translate(DefaultStorageRpc.java:202)
	at com.google.cloud.storage.spi.DefaultStorageRpc.get(DefaultStorageRpc.java:348)
	at com.google.cloud.storage.StorageImpl$5.call(StorageImpl.java:186)
	at com.google.cloud.storage.StorageImpl$5.call(StorageImpl.java:183)
	at com.google.cloud.RetryHelper.doRetry(RetryHelper.java:179)
	at com.google.cloud.RetryHelper.runWithRetries(RetryHelper.java:244)
	at com.google.cloud.storage.StorageImpl.get(StorageImpl.java:183)
	at com.google.cloud.storage.contrib.nio.CloudStorageFileSystemProvider.checkAccess(CloudStorageFileSystemProvider.java:531)
	at java.nio.file.Files.exists(Files.java:2385)
	at org.broadinstitute.hellbender.utils.io.IOUtils.assertFileIsReadable(IOUtils.java:567)
	at org.broadinstitute.hellbender.engine.FeatureDataSource.getFeatureReader(FeatureDataSource.java:248)
	at org.broadinstitute.hellbender.engine.FeatureDataSource.<init>(FeatureDataSource.java:209)
	at org.broadinstitute.hellbender.engine.VariantWalker.initializeDrivingVariants(VariantWalker.java:54)
	at org.broadinstitute.hellbender.engine.VariantWalkerBase.initializeFeatures(VariantWalkerBase.java:41)
	at org.broadinstitute.hellbender.engine.GATKTool.onStartup(GATKTool.java:432)
	at org.broadinstitute.hellbender.engine.VariantWalker.onStartup(VariantWalker.java:43)
	at org.broadinstitute.hellbender.cmdline.CommandLineProgram.runTool(CommandLineProgram.java:109)
	at org.broadinstitute.hellbender.cmdline.CommandLineProgram.instanceMainPostParseArgs(CommandLineProgram.java:169)
	at org.broadinstitute.hellbender.cmdline.CommandLineProgram.instanceMain(CommandLineProgram.java:188)
	at org.broadinstitute.hellbender.Main.runCommandLineProgram(Main.java:120)
	at org.broadinstitute.hellbender.Main.mainEntry(Main.java:141)
	at org.broadinstitute.hellbender.Main.main(Main.java:196)
Caused by: shaded.cloud-nio.com.google.api.client.http.HttpResponseException: 400 Bad Request
{
  "error" : "invalid_grant",
  "error_description" : "Token has been expired or revoked."
}
	at shaded.cloud-nio.com.google.api.client.http.HttpRequest.execute(HttpRequest.java:1061)
	at shaded.cloud-nio.com.google.auth.oauth2.UserCredentials.refreshAccessToken(UserCredentials.java:121)
	at shaded.cloud-nio.com.google.auth.oauth2.OAuth2Credentials.refresh(OAuth2Credentials.java:97)
	at shaded.cloud-nio.com.google.auth.oauth2.OAuth2Credentials.getRequestMetadata(OAuth2Credentials.java:74)
	at shaded.cloud-nio.com.google.auth.http.HttpCredentialsAdapter.initialize(HttpCredentialsAdapter.java:65)
	at com.google.cloud.HttpServiceOptions$1.initialize(HttpServiceOptions.java:233)
	at shaded.cloud-nio.com.google.api.client.http.HttpRequestFactory.buildRequest(HttpRequestFactory.java:93)
	at shaded.cloud-nio.com.google.api.client.googleapis.services.AbstractGoogleClientRequest.buildHttpRequest(AbstractGoogleClientRequest.java:300)
	at shaded.cloud-nio.com.google.api.client.googleapis.services.AbstractGoogleClientRequest.executeUnparsed(AbstractGoogleClientRequest.java:419)
	at shaded.cloud-nio.com.google.api.client.googleapis.services.AbstractGoogleClientRequest.executeUnparsed(AbstractGoogleClientRequest.java:352)
	at shaded.cloud-nio.com.google.api.client.googleapis.services.AbstractGoogleClientRequest.execute(AbstractGoogleClientRequest.java:469)
	at com.google.cloud.storage.spi.DefaultStorageRpc.get(DefaultStorageRpc.java:346)
	... 20 more
```

**Comment by droazen** [comment]
@jean-philippe-martin Any thoughts on what could be causing this?

**Comment by jean-philippe-martin** [comment]
Not yet. I need to reproduce the issue on my end and investigate.

**Comment by jean-philippe-martin** [comment]
I doubt this is the problem, but could you please confirm

 * Does the gsutil command work even when issued in the same terminal as the GATK command ?
 * If there is an index file (eg. .vcf.idx), could you please double-check that you can read it with gsutil as well?

In order to help me reproducing the issue,
 * Was this running from your desktop, or from a Google Compute (or Google Dataproc) instance?

**Comment by jean-philippe-martin** [comment]
I wasn't able to reproduce this exact issue. However at least I can make the error message a bit more user-friendly (submitting #2417 for review). Crucially this will now give the name of the file we're having issues with (thus removing the uncertainty about whether it's the data or the index file we cannot access).

**Comment by droazen** [comment]
@kcibul Can you try re-running with a build of https://github.com/broadinstitute/gatk/pull/2417, and paste the (hopefully) more detailed error message here?

Can you also try with a few different files in different buckets, and see whether you get the same error in every case?

**Comment by jean-philippe-martin** [comment]
gsutil working is not a good predictor of gatk working. It's possible for default credentials to be wrong but gsutil credentials to be fine at the same time. Here is an example of how to get into this situation:

```
// set application credentials
gcloud auth login
// unset default credentials (alternatively, forget to set them in the first place)
gcloud auth application-default revoke
// gsutil works
gsutil cat $VCF > /dev/null
// GATK does not work
./gatk-launch SelectVariants --verbosity=DEBUG -V $VCF -L 1:1000-2000 -O /tmp/foo.vcf
A USER ERROR has occurred: Couldn't read file (...)
```

Please make sure to set up Google Cloud access as follows:
```
$ gcloud auth application-default login
```

**Comment by kcibul** [comment]
Thanks for the ideas -- I found a few more details out this morning.

I was trying to use a service account, rather than my personal account, using `gcloud auth activate-service-account`.  This works for gcloud and gsutil commands, but doesn't seem to work with ADC very well evidently.  Once I changed and used my personal account via `gcloud auth application-default login` GATK4 no longer gave that error.

Then I found out that the file I gave you (which I picked because it's NA12878, but not where I originally found the problem) was not indexed.  So I went back to using the original file which has `foo.vcf.gz` as well as `foo.vcf.gz.tbi`.  GATK SelectVariants ran successfully.

Finally, I spun up a GCE-vm which is running with the service account I want, installed Java and GATK4 and was able to run the command successfully.

So it seems like the problem would be "how do I run using a service account from a non-GCE VM".  If there's an answer to that, that would be great.
